### PR TITLE
Fix: Prevent the daemon from starting if the sync box is not detected

### PR DESCRIPTION
**Pull request by esavary**
- At startup, the daemon checks if the trigger box is plugged in, throws an error and stops if this is not the case.
 - Add a check in the SOPs for the status of the daemon and precise the order in which to plug the USB cables.

### CODE REVIEW

File: code/synchronization/forward-trigger-service.py [hunk 1]
```diff
@@ -25,12 +25,13 @@
...
 import logging
-
 import janus
 import keyboard
+import usb.core
+import usb.util
 
 
-LISTEN = 2023
+LISTEN = 8888
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py]: Why this change?

File: docs/data-collection/pre-session.md [hunk 2]
```diff
@@ -191,6 +191,8 @@ Instructions of operations to be performed before the participant arrival, **bef
     !!! warning "If the monitor does not automatically switch the source of the screen, you can use the button below to switch it." 
     
         ![switch_screen](../assets/images/screen_switch.jpg)
+      
+- [ ] Connect the NEUROSPEC adapter (pink USB cable) to one of the USB port of the laptop *{{ secrets.hosts.psychopy | default("███") }}*. **This cable must always be connected last.**
```
Comment (oesteban) [inline docs/data-collection/pre-session.md]:
```suggestion
- [ ] Connect the MMBT-S Trigger Interface Box adapter (pink USB cable) to one of the USB ports of the laptop *{{ secrets.hosts.psychopy | default("███") }}*.
    ![neurospec](../assets/images/neurospec.jpg)
    
    !!! danger "The MMBT-S Trigger Interface MUST be connected to the laptop AFTER the trigger USB cable coming from the SyncBox. "
```

File: docs/data-collection/pre-session.md [hunk 3]
```diff
@@ -63,8 +63,8 @@ Instructions of operations to be performed before the participant arrival, **bef
     ![mac_setup](../assets/images/mac_setup.png) 
 - [ ] Connect one end of the parallel cable on the parallel plug on the back side of the STP100D unit of the BIOPAC and the other end to the parallel port of the NEUROSPEC parallel-to USB-converter.
     ![biopac-parallel-plug](../assets/images/biopac-parallel-plug.jpg "BIOPAC back side")
-    ![neurospec](../assets/images/neurospec.jpg)
-- [ ] Connect that NEUROSPEC adapter to one of the USB port of the laptop *{{ secrets.hosts.psychopy | default("███") }}*.
+    ![neurospec](../assets/images/neurospec.jpg) ** Do not connect the NEUROSPEC parallel-to USB-converter to *{{ secrets.hosts.psychopy | default("███") }}* now. It must always be connected after the USB cable of the sync box.**
+
```
Comment (oesteban) [inline docs/data-collection/pre-session.md]:
```suggestion
```

File: code/synchronization/forward-trigger-service.py (hunk 1 above)
Comment (esavary) [reply code/synchronization/forward-trigger-service.py]: I changed it back to 2023.

Comment (oesteban) [reply code/synchronization/forward-trigger-service.py]: The rationale is that you may have a jupyter notebook running and we don't want to have any conflicts.

File: code/synchronization/forward-trigger-service.py [hunk 4]
```diff
@@ -52,13 +53,10 @@ async def handle_client(
 ) -> None:
     """
     Handle client connections.
-
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py:55]: Please add these lines back, the convention of docstrings is that the first line is a one-line short description that starts with an action and ends in period. Then an empty line and then the long description may or may not expand through several lines.

BTW, I'm noticing that the parameters docstring is not numpy/napoleon compliant. I will fix that.

File: code/synchronization/forward-trigger-service.py [hunk 5]
```diff
@@ -52,13 +53,10 @@ async def handle_client(
...
     Read incoming data from the client and put it into the async queue.
-
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py:57]: Ditto

File: code/synchronization/forward-trigger-service.py [hunk 6]
```diff
@@ -52,13 +53,10 @@ async def handle_client(
...
     Parameters:
         reader (asyncio.StreamReader): The client's stream reader.
         async_q (janus.AsyncQueue[int]): The async queue for storing incoming signals.
-
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py:61]: Ditto (and so on, so forth all the way to the end)

File: docs/data-collection/pre-session.md [hunk 7]
```diff
@@ -200,6 +203,7 @@ Instructions of operations to be performed before the participant arrival, **bef
     
 - [ ] Configure the display settings of the laptop to mirror outputs and set a resolution of 800x600 for both screens.
 - [ ] Double check that the IP address corresponding to the ethernet interface of the *{{ secrets.hosts.psychopy | default("███") }}* laptop is correct. You can either run `ifconfig -a` or use the GUI. Make sure the IP/mask is **100.1.1.2/24**, and the protocol is IP version 4. Execute `ping 100.1.1.1` to see if the ET is responding to echoes.
+- [ ] Check that the service to synchronize the triggers is on with `sudo systemctl status forward-trigger`. If not, run `sudo systemctl status forward-trigger` and recheck the status.
```
Comment (oesteban) [inline docs/data-collection/pre-session.md]:
```suggestion
- [ ] Check that the service to synchronize the triggers is up with `sudo systemctl status forward-trigger`.

    !!! warning "If the service is down, manually force its start"
    
        - [ ] Run `sudo systemctl status forward-trigger`
        - [ ] Recheck the status with `sudo systemctl status forward-trigger`.
    
    !!! important "These commands are executed with `sudo`"
    
        The console will prompt you for the common user password: `{{ secrets.login.password_hos68752 | default("****") }}`

```

File: docs/data-collection/pre-session.md [hunk 8]
```diff
@@ -63,8 +63,6 @@ Instructions of operations to be performed before the participant arrival, **bef
     ![mac_setup](../assets/images/mac_setup.png) 
 - [ ] Connect one end of the parallel cable on the parallel plug on the back side of the STP100D unit of the BIOPAC and the other end to the parallel port of the NEUROSPEC parallel-to USB-converter.
```
Comment (oesteban) [inline docs/data-collection/pre-session.md]: Please revise this (I can't using the PR review utility because you did not modify this line):

1. Let's be more precise and replace any instance of "NEUROSPEC <blah blah>" with "MMBT-S Trigger Interface Box adapter" or "MMBT-S Trigger Interface Box adapter (N-shaped pink box)".
2. Let's be more precise and say that we want to connect the parallel cable to the 25-pin socket at the back of the SPT100D.

File: code/synchronization/forward-trigger-service.py [hunk 9]
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
...
     """
     Main function.
-
     Run the signal server and forward the signals to the serial port.
-
     """
+    # Check if the USB device is connected
+    initial_usb_connection = check_usb_device_connected(
+        "07c0", "0101"
+    )  # for testing "0bc2", "2322"
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py]: please move these magic numbers to the top of the file, by setting some static-ish variables (python doesn't have that):

```Python
USB_MMBTS_DEVICE_ID = ("07c0", "0101")
"""Device ID of the Neurospec's MMBT-S trigger adaptor."""
USB_TESTING_DEVICE_ID = ("0bc2", "2322")
"""Device ID for testing purposes."""
```

and then
```suggestion
    ensure_usb_device_connected(*USB_MMBTS_DEVICE_ID)
```

File: code/synchronization/forward-trigger-service.py [hunk 10]
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
...
 
 
+def check_usb_device_connected(usb_vendor_id, usb_product_id):
+    """
+    Check if a USB device with the specified vendor and product IDs is connected.
+    Parameters:
+        usb_vendor_id (str): The USB vendor ID.
+        usb_product_id (str): The USB product ID.
+    Returns:
+        bool: True if the device is connected, False otherwise.
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py]:
```suggestion
    Check if a USB device with the specified vendor and product IDs is connected.
    
    Parameters
    ----------
    usb_vendor_id : :obj:`str`
        Vendor ID of a USB device.
    usb_product_id : :obj:`str`
        Product ID of a USB device.

    Returns
    -------
    connected : :obj:`bool`
        ``True`` when the device is connected. If the device is not connected, this
        function raises a :obj:`RuntimeError`.

```

File: code/synchronization/forward-trigger-service.py [hunk 11]
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
...
     Parameters:
         sync_q (janus.SyncQueue[int]): The sync queue for triggering the signal.
-
     """
     sync_q.put(b"\x01")
     sync_q.join()
     logging.info("Scanner trigger received")
 
 
+def check_usb_device_connected(usb_vendor_id, usb_product_id):
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py]: Since this function raises exception if not connected, perhaps it would be better to call it differently (this is just a suggestion):

```suggestion
def ensure_usb_device_connected(usb_vendor_id, usb_product_id):
```

File: code/synchronization/forward-trigger-service.py [hunk 12]
```diff
@@ -127,25 +119,43 @@ async def forward_signals(serial_port: str, async_q: janus.AsyncQueue[int]) -> N
...
+
 async def main() -> None:
     """
     Main function.
-
     Run the signal server and forward the signals to the serial port.
-
     """
+    # Check if the USB device is connected
+    initial_usb_connection = check_usb_device_connected(
```
Comment (oesteban) [inline code/synchronization/forward-trigger-service.py]: Why keeping a variable if this is expected to crash if the device is not present?

```suggestion
    ensure_usb_device_connected(
```

**Review by oesteban — COMMENTED** [review]
this is starting to look great! thanks for this :)

**Comment by esavary** [comment]
Thanks for the review! I've integrated the comments in the last commits.

**Comment by oesteban** [comment]
@esavary I'm merging because the code looks good and I'm sure you have run `mkdocs serve` and have checked that the markdown renders correctly :)
